File: src/main/java/org/jetbrains/fortran/lang/parser/FortranFixedFormParserDefinition.java
package org.jetbrains.fortran.lang.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.fortran.lang.FortranTypes;
import org.jetbrains.fortran.lang.core.stubs.FortranFixedFormFileStub;
import org.jetbrains.fortran.lang.lexer.FortranLexer;
import org.jetbrains.fortran.lang.psi.FortranFixedFormFile;

import static org.jetbrains.fortran.lang.psi.FortranTokenType.*;

public class FortranFixedFormParserDefinition implements ParserDefinition {
    @NotNull
    @Override
    public Lexer createLexer(Project project) { return new FortranLexer(true); }

    @Override
    public PsiParser createParser(Project project) {
        return new FortranParser();
    }

    @Override
    public IFileElementType getFileNodeType() {
        return FortranFixedFormFileStub.Type.INSTANCE;
    }

    @NotNull
    @Override
    public TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }

    @NotNull
    @Override
    public TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @NotNull
    @Override
    public TokenSet getStringLiteralElements() {
        return STRINGS;
    }

    @NotNull
    @Override
    public PsiElement createElement(ASTNode astNode) {
        PsiElement element = FortranManualPsiElementFactory.createElement(astNode);
        return element == null ? FortranTypes.Factory.createElement(astNode) : element;
    }

    @Override
    public PsiFile createFile(FileViewProvider fileViewProvider) {
        return new FortranFixedFormFile(fileViewProvider);
    }

    @Override
    public SpaceRequirements spaceExistanceTypeBetweenTokens(ASTNode astNode, ASTNode astNode1) {
        return SpaceRequirements.MAY;
    }
}

File: src/main/kotlin/org/jetbrains/fortran/lang/parser/FortranFixedFormParserDefinition.kt
package org.jetbrains.fortran.lang.parser

import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet
import org.jetbrains.fortran.lang.FortranTypes
import org.jetbrains.fortran.lang.lexer.FortranLexer
import org.jetbrains.fortran.lang.psi.FortranFixedFormFile
import org.jetbrains.fortran.lang.psi.FortranTokenSets
import org.jetbrains.fortran.lang.stubs.FortranFixedFormFileStub

class FortranFixedFormParserDefinition : ParserDefinition {
    override fun createLexer(project: Project): Lexer {
        return FortranLexer(true)
    }

    override fun createParser(project: Project): PsiParser {
        return FortranParser()
    }

    override fun getFileNodeType(): IFileElementType {
        return FortranFixedFormFileStub.Type
    }

    override fun getWhitespaceTokens(): TokenSet {
        return FortranTokenSets.WHITE_SPACES
    }

    override fun getCommentTokens(): TokenSet {
        return FortranTokenSets.COMMENTS
    }

    override fun getStringLiteralElements(): TokenSet {
        return FortranTokenSets.STRINGS
    }

    override fun createElement(astNode: ASTNode): PsiElement {
        val element = FortranManualPsiElementFactory.createElement(astNode)
        return element ?: FortranTypes.Factory.createElement(astNode)
    }

    override fun createFile(fileViewProvider: FileViewProvider): PsiFile {
        return FortranFixedFormFile(fileViewProvider)
    }

    override fun spaceExistenceTypeBetweenTokens(astNode: ASTNode, astNode1: ASTNode): ParserDefinition.SpaceRequirements {
        return ParserDefinition.SpaceRequirements.MAY
    }
}